working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+potentially nullable.
+
+It is not an error for the body of a `lazy` field to reference `this`.
+
+It is an error for a toplevel variable, a static field, or a formal parameter to
+be declared `lazy`.
+
+It is an error if the type `T` in the **on-catch** clause `on T catch` is
+potentially nullable.
+
+### Assignability
+
+The definition of assignability is changed as follows.  
+
+A type `T` is **assignable** to a type `S` if `T` is `dynamic`, or if `S` is a
+subtype of `T`.
+
+### Generics
+
+The default bound of generic type parameters is treated as `Object?`.
+
+### Type promotion, Definite Assignment, and Definite Completion
+
+**TODO** Fill this out.
+
+### Null promotion
+
+The machinery of type promotion is extended to promote the type of variables
+based on nullability checks subject to the same set of restrictions as normal
+promotion.  The relevant checks and the types they are considered to promote to
+are as follows.
+
+A check of the form `e == null` or of the form `e is Null` where `e` has static
+type `T` promotes the type of `e` to `Null` in the `true` continuation, and to
+**NonNull**(`T`) in the
+`false` continuation.
+
+A check of the form `e != null` or of the form `e is T` where `e` has static
+type `T?` promotes the type of `e` to `T` in the `true` continuation, and to
+`Null` in the `false` continuation.
+
+The static type of an expression `e!` is **NonNull**(`T`) where `T` is the
+static type of `e`.
+
+The **NonNull** function defines the null-promoted version of a type, and is
+defined as follows.
+
+- **NonNull**(Null) = Never
+- **NonNull**(_C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>) = _C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>  for class *C* other than Null
+- **NonNull**(FutureOr<_T_>) = FutureOr<_T_>   
+- **NonNull**(_T_<sub>0</sub> Function(...)) = _T_<sub>0</sub> Function(...)
+- **NonNull**(Function) = Function
+- **NonNull**(Object) = Object
+- **NonNull**(Never) = Never
+- **NonNull**(dynamic) = dynamic
+- **NonNull**(void) = void   
+- **NonNull**(_X_) = X & **NonNull**(B), where B is the bound of X.
+- **NonNull**(_X_ & T) = X & **NonNull**(T)
+- **NonNull**(_T_?) = **NonNull**(_T_)
+- **NonNull**(_T_\*) = **NonNull**(_T_)
+
+#### Type promotion, Definite Assignment, and Definite Completion
+
+These are extended as per separate proposal.
+
+### Runtime semantics
+
+#### Null assertion operator
+
+An expression of the form `e!` evaluates `e` to a value `v`, throws a runtime
+error if `v` is `null`, and otherwise evaluates to `v`.
+
+#### Null aware operator
+
+An expression of the form `e?.<tail>` where `<tail>` is any sequence of
+selectors evaluates to `null` if `e` evaluates to `null`, and otherwise
+evaluates to the same value as `e.<tail>`.
+
+An expression of the form `e?..<tail>` where `<tail>` is any sequence of
+selectors evaluates to `null` if `e` evaluates to `null`, and otherwise
+evaluates to the same value as `e..<tail>`.
```

Comment:
Need something more to handle `x?..foo()..bar()`. This should elide both calls.

Reply:
Updated with a more precise specification, PTAL.